lib/vm/src/memory.rs

```diff
@@ -46,28 +46,40 @@ pub fn read_region(ctx: &Ctx, ptr: u32) -> Vec<u8> {
     }
 }
 
-// Returns how many bytes written on success
-// negative result is how many bytes requested if too small
-pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> i32 {
+/// A prepared and sufficiently large memory Region is expected at ptr that points to pre-allocated memory.
+///
+/// Returns number of bytes written on success.
+pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> Result<usize, Error> {
     let region = to_region(ctx, ptr);
     if data.len() > (region.len as usize) {
-        return -(data.len() as i32);
+        return ExecutionErr {
```

Comment:
Okay. Now I see where this error comes from. And not using negative numbers to signal errors is good (and not so C-like).

However... I wonder if you change the return value of `c_read` on failure: https://github.com/confio/cosmwasm/blob/master/src/imports.rs#L54-L60

Actually implementing dynamic buffer-sizes is a good TODO as well. So we load data into eg 1K preallocated, which works 99.8% of the time. But if someone does store 4k blob in the storage, we can just prepare a larger Buffer and try the call again.

Reply:
Arg, I was not aware that the actual return value of `c_read` mattered because it was not used. In https://github.com/confio/cosmwasm/pull/118/commits/8662c33fee8db9bdf4d02afa2b3805109387f60d I started documenting the imports, treating all negative values as errors.

Two solutions come to my mind:
1. Create a single return code for _buffer too small_
2. Do some crazy bit operation, encoding both _buffer too small_ and the required size into the return value

I'm in favour of 1, which is already better than what we need in `if read < 0 {`.

Reply:
What about this as a mid term solution:
1. Document 16 KiB as the max storage entry size supported by CosmWasm
2. Create and document something like `static BUFFER_TOO_SMALL_ERROR: i32 = -1002;`
3. Let contracts retry with `2^i` KiB for `i = 0, 1, 2, 3, 4` (i.e. 1, 2, 4, 8, 16 KiB).

Long term I'd aim for an API that always succeeds in the first attempt.

Reply:
Ah, even better. Since the contract allocates the memory, the contract itself can make reasonable assumptions on which data size to expect. This can be a trial and error approach or something more sophisticated to safe gas.

In this case, a _buffer too small_ error seems like sufficient information as an error.

Reply:
I mean, the info that "too small, you need 4871 bytes" seems nicer in a way. But so far anyone over 1k is an edge case.

I don't need that all to work perfect now, I just don't want to make it hard to add that in a good way later.